go/multiorch/recovery/healthcheck.go

```diff
@@ -0,0 +1,322 @@
+// Copyright 2025 Supabase, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package recovery
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/metric"
+
+	"github.com/multigres/multigres/go/clustermetadata/topo"
+	"github.com/multigres/multigres/go/multiorch/store"
+	"github.com/multigres/multigres/go/pb/clustermetadata"
+)
+
+// pollPooler performs health check on a single pooler instance.
+// This is analogous to Vitess's DiscoverInstance.
+//
+// It includes:
+// - Deduplication via cache (prevents redundant checks)
+// - Latency tracking
+// - Warnings if check exceeds poll interval
+// - Detailed metrics for observability
+//
+// The function calls either GetPrimaryStatus or GetReplicaStatus via gRPC
+// depending on the pooler type, then updates the in-memory store.
+//
+// Parameters:
+//   - poolerID: The pooler's ID
+//   - pooler: The pooler's health info from the store
+//   - forceDiscovery: If true, bypass cache and up-to-date checks (force poll)
+func (re *Engine) pollPooler(poolerID *clustermetadata.ID, pooler *store.PoolerHealth, forceDiscovery bool) {
+	// Skip if this pooler is marked as forgotten (shouldn't happen, but defensive)
+	if pooler == nil || pooler.MultiPooler == nil {
+		re.logger.Debug("skipping poll of nil pooler", "pooler_id", topo.MultiPoolerIDString(poolerID))
+		return
+	}
+
+	poolerIDStr := topo.MultiPoolerIDString(poolerID)
```

Comment:
nit: This can happen before the previous line, then poolerIDStr can used in it too